### ISSUE TITLE
incomplete MerkleProof for some StorageValues

### ISSUE THREAD

**Issue by simon-jentzsch** [NethermindEth/nethermind#1461]
If I ask for proof for a storage value sometimes I get the following:

```js
{
  "id":1,
  "jsonrpc":"2.0",
  "method":"eth_getProof",
  "params":[
     "0xa14631f8867071320e5a3fc432f0f7165d77bd56", 
     ["0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e58a"],
    "latest"
   ]
}
```

I get the following response:

```js
{
  "accountProof": [
    "0xf90211a07ea67a5a5c827e7ea66f62f84b8ef724da250a0ca14a69053ac53df5518798a3a0a68559fc085526ffc0df224ede7bad7856bd9158ce80a99ec5cdb31537521aa6a0dbc952a2d35e415feec1140b489fefb8a12f741496b5ad52fa5d4050e9d8194da0d722b5dbf4e93c38c60184aebdd43a54fa30b5b4df6372141e3dbf28b89f1430a094c1967d16e5fa617d649c2b463a99dd62eb6f96d1a0416cf3ba7f5d94b1a00aa0fcace36b14ac3768bb4bbabc0537d3985db87a98063b73f518e018b5af604872a0b4b3a4f1b25e8f1ce2046da390c32d0e724ef5191f7afe404b0e31df676bc962a08c4baa71c4196361d1ffbf65f983dbf2ef201eeea3728ed84bf09dff29ecc0ffa011eab12a570b4327db9561aac32c310cc1fcdda4d16cf7400e18fafb93505de3a07e86ab91829f1b5dad1208b083a246b46c3bd415adb69d696161c90f0ea3cf26a069e074d75e07f9280dbc7c3b056fde57ff80e07a26796294548818b838e47967a095d81a13562435be3f0aadae168dfb664a4d803538c14cb8f68791935956d781a0c58751a700f2aa2728597eb8207c92cde04d76a862697c792c5f3e9880cb5af1a00e616e523931633c4277e080cec788a9120c116121047e8001f6ad923ea872eda0ce3b5eec426cf8afcd4cc5bf6ec44d899b381436afcfbc2ff98616bd3377dbe9a0c862f705f14cc1ddf8aea388c35b137f78f1f4e9a42455f6d3b1a81677bf0dfe80",
    "0xf90211a02177c70b3040ad4ecf4dcb38a53b088eb4a5873d7600864358511ec71000f0c1a065c74138737689f1543ae7729c56fe6307b1628d5066dc6f4613e38b59a99a13a03af4b525fe53b5616fdcc26cad609fad70642bda14199fd0b8fee67ad958a582a06a49756b83fc1bd2fc98a444405eed63879dc0bf912afa8ad7b10b43c869d3c9a0dbac7ab4421c95c30a8578e9e543f559809eebdb421295672aea84d82eb31f1aa0f89bc179f4ca68680870c6524897a26ce435fc5644fbca81a8cdabddd2c96724a0bae3fb5143229f589180001c10f02543b17f668e09dcf51bf9cf94e7f5828c19a0122adeb25d5151dd6fb3c8e81c7430412648ed684d1dfc90866e7f3fa186d26aa00bb55f351e199df83a7c88d76b7ea48a290fc7a0608fe2a75a760bed512153eea0e0da4600175f18c463f17356bf9ccf5077a05e1ca14d061a8b01cdd99d477682a02ec7b729c81501f645f6daadc75cb6721a3a9248c4955ec4c90ceb0f9e1e7c2aa0e4bfd6520befbbb5181c465c040c1124d2650b4e8b0e78c5964c6f146741f493a0141ea0bef4d35f4f8444460756373b46d0d976397ca8fe1b9b894dd7dc71372ea0f87a732c13fcf50d5f3eb7a7e06fbe050537a5015b379da96c91c6670aff4fd8a00960616005c80ef77df7c1872ff6f09ea0e7c3c8848178be070dd09a028180daa0eec93bfef36234842a191c3b7879931559162bffd823e387f446e5856a15fde380",
    "0xf90131a0a884e21ad0b2ac21cea19157eff89ed9b5b17c356763341c01552e82195357a8a03af33dc4a744ab5edd42c5c5fdc99e424fde379ac9650ffb1246d28945dee78aa0c6761c20b2a2b12c01561db76a46c95e768a6c20908f08206d65766ab570bee780a09642d8ce2ba67d9897ab525248d6f36a82862e057b4675c5d84c3a956e0fed92808080a004a0eafc3cdb494daeb274245c78aca2260ede32cac699064e95df455dcbb05b8080a0f56d396c002a2a43be0544be4e5bca4290dde283aea26b25a408117df7405c9da077041e5a6b0816354432ea5a68b34b58a9cfd967f1bb4c27cedffc0b753d0b88a06c23fa43987fe4cc5d35795d499a93676182f023822591b27160f94380661daa80a0dcac789e8cadd3bb1405bf3e76355b4e8f5bf4d9b0d991c5d9e1f2be113d4d2280",
    "0xf8689f33089359897e301d75ff74467785ee917b0f2f0326800a9ef12b4f5c28172fb846f8440180a046ae33d3e8eb4b8d3fb431172b03f979e3d1d1096fae61bc461efaa137f954d2a0f081344097c1981565c77bf3af4b0f9ee3c194343e4fd8fcabdc8d8f35810285"
  ],
  "address": "0xa14631f8867071320e5a3fc432f0f7165d77bd56",
  "balance": "0x0",
  "codeHash": "0xf081344097c1981565c77bf3af4b0f9ee3c194343e4fd8fcabdc8d8f35810285",
  "nonce": "0x1",
  "storageHash": "0x46ae33d3e8eb4b8d3fb431172b03f979e3d1d1096fae61bc461efaa137f954d2",
  "storageProof": [
    {
      "key": "0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e58a",
      "proof": [
        "0xf90211a06fb9269e5ce0e5e192d730d85010581ebc68d2bd29f309406ed9a765f319beb5a04a0b3a20d9bfa9118b00cb4fd87b526558495204077942c7e44538b5a0003b83a02ca7736ac27674e9a6f52e33b522b835fc19a24ce409217717f0ec50c7390594a0415094b03547600e43c2945fcaa89d973be325fc7c4908eb52a595b4c51dfb60a090ec0e30643b79a186b3b1da3fff9df87ac5010ce1e59392b5b1a97b80d05662a03844ebca032a18ae9f58a9d6c8e3cf4cd101b7057cceba4d456339edc266a945a07934a73a0846884e5e5d68ac5c9d5e8cbb12128284e9fd07e4a2a0c2f0b23e00a002dd6cdd5b0754831d2132f6e37e74bea5875e3f44b76f63888db9df5eb597cfa0a1eef492150d4432fa593dc927ccebe40b5f8e560f9c270e4f09cb1820c8a9a7a084e298383b35d9f1b7a9c82a8c098ff6400ff2594da5d2c5e6a3710a10f8b0b1a0b62c474a8fdd75de4910d635d10cde047ddcc83a1cb0bcd7594492162597548aa044127bad70938219844bfc537dd6e10518a5fe54d79e94541c3d74d936757b83a0013f438998b6555d7ac86171593b67ffba61517ea43f95f63ff5328157231b92a0319fbff13175ee3b6d52bdf81f39b164d8ff973a81ce8c4e1164acd85c2d49d3a0cca1bd471dee7e1e6f7c5077ae976daaaee628679ea75a7f06d5f4b4fbcbd3d2a02d7d70fa3fead467fbc99c778b27469d6502b3d484fec8be1c26167622c4cc6380",
        "0xf8f1a07b9d6e20503402214ffa2af2583229bf2c2ccf0e9a158b63fa8620abf99e3c20a0413e763c37f5ff3eed09c3974a8fc0dbc894101a737de33fdc08f0db1dc5f671a04fe7ed3025f88ccae2a8eba26adacba3751882939f55611ebe600b163cef99188080a00266783745904038f85e2621a6b49f0b1b713c52cea5a6877e7e9b291c805c39a0e0534a7d8ff75d3de5ece315f72b5b7b0408d8325381fdd2fd7dfebf17bab45d808080a0e8fbcddb534b5f45a9b7e50fc053e8948679432b25810d95c658e2983cd4dfa6a060e2123ee086d2c82e81a8626616b8d6bd74b5ab0d62ed10780b07125daafc1e8080808080",
        "0xf851808080a0bc62076eabe36d3a74444ba9adeab24dcfffe622dd8c0b480e5ad7d9c53d6aff80808080808080a0206cddcbd37b1c640e703cd24ad0749730593a017dd156af5fc12ae121b3eb8e8080808080",
        "0xe216a037ade216bad5f991902d9d0c6da7dec03bdd0313c456b2098634ed993d8e89d0"
      ],
      "value": null
    }
  ]
}
```

1. the value should never be null. For a non existing value it should be `0x000000000...`
2. But In this case the value exists! (If I call `eth_getStorageAt` I get  `0xe88116cd7cc1c8d423c4f1eaafb5ee251f09485aac478313d7a965c5bde1e089` as result)

I also tried to follow the merkle proof.
And this merkle proof translate into 3 correct branch-nodes and then ends with a Extension-Node (which has a correct path), but the proof seems to be incomplete.
It looks like after the extension-node a branch or leaf-node is missing.

## Analysis:

path or nibbles (`keccak('0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e58a')`) =
`8ab6fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35`

1. node (root) : PATH **8** ab6fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35
```
[ MERKLE BRANCH  Hash : 0x46ae33d3e8eb4b8d3fb431172b03f979e3d1d1096fae61bc461efaa137f954d2
   0                    : <hash>           0x6fb9269e5ce0e5e192d730d85010581ebc68d2bd29f309406ed9a765f319beb5
   1                    : <hash>           0x4a0b3a20d9bfa9118b00cb4fd87b526558495204077942c7e44538b5a0003b83
   2                    : <hash>           0x2ca7736ac27674e9a6f52e33b522b835fc19a24ce409217717f0ec50c7390594
   3                    : <hash>           0x415094b03547600e43c2945fcaa89d973be325fc7c4908eb52a595b4c51dfb60
   4                    : <hash>           0x90ec0e30643b79a186b3b1da3fff9df87ac5010ce1e59392b5b1a97b80d05662
   5                    : <hash>           0x3844ebca032a18ae9f58a9d6c8e3cf4cd101b7057cceba4d456339edc266a945
   6                    : <hash>           0x7934a73a0846884e5e5d68ac5c9d5e8cbb12128284e9fd07e4a2a0c2f0b23e00
   7                    : <hash>           0x02dd6cdd5b0754831d2132f6e37e74bea5875e3f44b76f63888db9df5eb597cf
   8                    : <hash>           0xa1eef492150d4432fa593dc927ccebe40b5f8e560f9c270e4f09cb1820c8a9a7
   9                    : <hash>           0x84e298383b35d9f1b7a9c82a8c098ff6400ff2594da5d2c5e6a3710a10f8b0b1
   A                    : <hash>           0xb62c474a8fdd75de4910d635d10cde047ddcc83a1cb0bcd7594492162597548a
   B                    : <hash>           0x44127bad70938219844bfc537dd6e10518a5fe54d79e94541c3d74d936757b83
   C                    : <hash>           0x013f438998b6555d7ac86171593b67ffba61517ea43f95f63ff5328157231b92
   D                    : <hash>           0x319fbff13175ee3b6d52bdf81f39b164d8ff973a81ce8c4e1164acd85c2d49d3
   E                    : <hash>           0xcca1bd471dee7e1e6f7c5077ae976daaaee628679ea75a7f06d5f4b4fbcbd3d2
   F                    : <hash>           0x2d7d70fa3fead467fbc99c778b27469d6502b3d484fec8be1c26167622c4cc63
   value                : 0                <EMPTY>
]
```
2. node  PATH **8a** b6fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35
```
[ MERKLE BRANCH  Hash : 0xa1eef492150d4432fa593dc927ccebe40b5f8e560f9c270e4f09cb1820c8a9a7
   0                    : <hash>           0x7b9d6e20503402214ffa2af2583229bf2c2ccf0e9a158b63fa8620abf99e3c20
   1                    : <hash>           0x413e763c37f5ff3eed09c3974a8fc0dbc894101a737de33fdc08f0db1dc5f671
   2                    : <hash>           0x4fe7ed3025f88ccae2a8eba26adacba3751882939f55611ebe600b163cef9918
   3                    : 0                <EMPTY>
   4                    : 0                <EMPTY>
   5                    : <hash>           0x0266783745904038f85e2621a6b49f0b1b713c52cea5a6877e7e9b291c805c39
   6                    : <hash>           0xe0534a7d8ff75d3de5ece315f72b5b7b0408d8325381fdd2fd7dfebf17bab45d
   7                    : 0                <EMPTY>
   8                    : 0                <EMPTY>
   9                    : 0                <EMPTY>
   A                    : <hash>           0xe8fbcddb534b5f45a9b7e50fc053e8948679432b25810d95c658e2983cd4dfa6
   B                    : <hash>           0x60e2123ee086d2c82e81a8626616b8d6bd74b5ab0d62ed10780b07125daafc1e
   C                    : 0                <EMPTY>
   D                    : 0                <EMPTY>
   E                    : 0                <EMPTY>
   F                    : 0                <EMPTY>
   value                : 0                <EMPTY>
]
```
3. node  PATH **8ab** 6fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35
```
[ MERKLE BRANCH  Hash : 0xe8fbcddb534b5f45a9b7e50fc053e8948679432b25810d95c658e2983cd4dfa6
   0                    : 0                <EMPTY>
   1                    : 0                <EMPTY>
   2                    : 0                <EMPTY>
   3                    : <hash>           0xbc62076eabe36d3a74444ba9adeab24dcfffe622dd8c0b480e5ad7d9c53d6aff
   4                    : 0                <EMPTY>
   5                    : 0                <EMPTY>
   6                    : 0                <EMPTY>
   7                    : 0                <EMPTY>
   8                    : 0                <EMPTY>
   9                    : 0                <EMPTY>
   A                    : 0                <EMPTY>
   B                    : <hash>           0x206cddcbd37b1c640e703cd24ad0749730593a017dd156af5fc12ae121b3eb8e
   C                    : 0                <EMPTY>
   D                    : 0                <EMPTY>
   E                    : 0                <EMPTY>
   F                    : 0                <EMPTY>
   value                : 0                <EMPTY>
]
```
4. node  PATH **8ab6** fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35
```
[ MERKLE EXTENSION  Hash : 0x206cddcbd37b1c640e703cd24ad0749730593a017dd156af5fc12ae121b3eb8e
   path                 : EXTENSION (odd)  0x1 6
   value                : <hash>           0x37ade216bad5f991902d9d0c6da7dec03bdd0313c456b2098634ed993d8e89d0
]
```

So we are missing the rest of the path : `fad336e69b72e50a4cec1eff1aac27f3fe4a9e50ec468c2897f8dff1de35`, which should point to a node with hash `0x37ade216bad5f991902d9d0c6da7dec03bdd0313c456b2098634ed993d8e89d0`

**Comment by tkstanczak** [comment]
it was not following Extensions properly here

built a mini-fuzzer and fixed

**Comment by tkstanczak** [comment]
fix released in 1.6.8
